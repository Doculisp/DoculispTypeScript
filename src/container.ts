import { IContainer, IDependencyContainer, IDependencyManager, IDictionary, IRegisterable, ITestableContainer, Valid } from "./types.containers";
import { globSync } from 'glob';

function findModules(container: IContainer) {
    let moduleNames = (
        globSync('./dist/**/*.js')
        .map(p => {
            return p.replace('dist', '.').replace('\\src', '').replace('.js', '').replaceAll('\\', '/');
        })
    );

    moduleNames.forEach(n => {
        import(n).then(m => {
            const keys = Object.keys(m);
            keys.
                filter(key => {
                    let value = m[key];
                    return !!value.name && !!value.builder;
                }).
                forEach(key => container.register(m[key]));
        });
    });
}

class Container implements ITestableContainer {
    _registry: IDictionary<IRegisterable> = {};
    _replacements: IDictionary<IRegisterable> = {};
    _cache: IDictionary<Valid<any>> = {};
    _repCache: IDictionary<Valid<any>> = {};
    _testable = false;

    constructor();
    constructor(registry: IDictionary<IRegisterable>);
    constructor(registry?: IDictionary<IRegisterable>) {
        this._testable = !!registry;
        if(!!registry) {
            const keys = Object.keys(registry);
            keys.forEach(key => this._registry[key] = registry[key]);
        }
        else {
            findModules(this);
        }
    }

    _hasKey(registry: IDictionary<IRegisterable>, key: string): boolean {
        return !!registry[key];
    }

    _register(registry: IDictionary<IRegisterable>, registerable: IRegisterable): ITestableContainer {
        let key = registerable.name;
        if (this._hasKey(registry, key)) {
            throw new Error(`Module named "${key}" already registered.`);
        }
        registry[key] = registerable;
        return this
    }

    register(registerable: IRegisterable): ITestableContainer {
        if (!registerable.name) {
            throw new Error('Must have a valid name to be registered.');
        }
        this._register(this._registry, registerable);
        return this;
    }
    
    registerValue(value: any, name?: string | undefined): ITestableContainer {
        if (!value.name && !name) {
            throw new Error('Most provide a name as a property or a parameter when registering a value as a module.');
        }

        const registerable: IRegisterable = {
            builder: () => { return value; },
            name: name ?? value.name,
            singleton: true
        };

        return this.register(registerable);
    }
    
    registerBuilder(fn: (...args: any[]) => any, dependencies: string[], name?: string | undefined, singleton?: boolean | undefined): ITestableContainer {
        if(!fn.name && !name) {
            throw new Error('Must provide a name on the function or as a parameter to register a builder.');
        }

        const registerable: IRegisterable = {
            builder: fn,
            name: name ?? fn.name,
            dependencies: dependencies ?? [],
            singleton: !!singleton,
        }
        
        this.register(registerable);
        return this;
    }

    _build<T>(moduleName: string, modules: string[]): Valid<T> {
        let moduleBuilder = this._replacements[moduleName];
        let replaced = true;
        if (!moduleBuilder) {
            moduleBuilder = this._registry[moduleName];
            replaced = false;
        }

        if (!moduleBuilder) {
            let t: any = undefined;
            try {
                t = require(moduleName);
            }
            catch {
            }

            if (!t) {
                throw new Error(`No module called "${moduleName}" registered`);
            }
            this.registerBuilder(() => require(moduleName), [], moduleName);
            moduleBuilder = this._registry[moduleName] as IRegisterable;
        }

        if(this._repCache[moduleName]) {
            return this._repCache[moduleName];
        }
        
        if(!replaced && this._cache[moduleName]) {
            return this._cache[moduleName];
        }

        let recursive = 
            moduleBuilder.dependencies?.filter(name => modules.includes(name)) ?? [];

        modules[modules.length] = moduleName;

        if (0 < recursive.length) {
            let moduleNames = modules.map(name => `"${name}"`).join(' => ');
            let recursiveModules = recursive.map(name => `"${name}"`).join(', ');
            throw new Error(`Circular dependencies between (${moduleNames} => [${recursiveModules}])`);
        }

        let dependencies: Valid<any>[] = 
            moduleBuilder.dependencies?.map((name) => this._build(name, modules)) ?? [];

        let result: Valid<T> = moduleBuilder.builder.apply(moduleBuilder.builder, dependencies);

        if(moduleBuilder.singleton) {
            if(replaced) {
                this._repCache[moduleName] = result;
            }
            else {
                this._cache[moduleName] = result;
            }
        }

        return result;
    }
    
    build(moduleName: string): Valid<any> {
        return this._build<any>(moduleName, []);
    }
    
    buildAs<T>(moduleName: string): Valid<T> {
        return this._build<T>(moduleName, []);
    }
    
    buildTestable(): ITestableContainer {
        return new Container(this._registry);
    }
    
    supportsReplace(): boolean {
        return this._testable;
    }

    _replace(registerable: IRegisterable, isPackage?: boolean) : ITestableContainer {
        let key = registerable.name;
        if(!isPackage && !this._hasKey(this._registry, key)) {
            throw new Error(`Cannot replace module "${key}" as it has not been registered.`);
        }
        if(!isPackage && this._hasKey(this._replacements, key)) {
            throw new Error (`Cannot replace module "${key}" as it has not been registered.`)
        }

        return this._register(this._replacements, registerable);
    }
    
    replace(registerable: IRegisterable): ITestableContainer {
        return this._replace(registerable);
    }
    
    replaceBuilder(fn: (...args: any[]) => any, dependencies: string[], name?: string | undefined, singleton?: boolean | undefined): ITestableContainer {
        if (!fn.name && !name) {
            throw new Error('Must provide a name either on the passed method or parameter to replace builder.');
        }

        let registerable: IRegisterable = {
            builder: fn,
            dependencies: dependencies ?? [],
            name: name ?? fn.name,
            singleton: !!singleton,
        };
        
        return this._replace(registerable);// 
    }
    
    replaceValue(value: any, name?: string | undefined): ITestableContainer {
        let key = name ?? value.name;
        if (!key) {
            throw new Error('Cannot replace value unless it has a name property or the name is passed as a parameter.');
        }

        return this.replaceBuilder(() => { return value; }, [], key, true);
    }
    
    replacePackageBuilder(fn: (...args: any[]) => any, name?: string | undefined, singleton?: boolean | undefined): ITestableContainer {
        if(!fn.name && !name) {
            throw new Error('Must provide a name either on the passed method or parameter to replace package builder.');
        }

        let registerable: IRegisterable = {
            builder: fn,
            name: name ?? fn.name,
            singleton: !!singleton,
        };
        
        return this._replace(registerable, true);
    }
    
    replacePackageValue(value: any, name?: string | undefined): ITestableContainer {
        if(!value.name && !name) {
            throw new Error('Cannot replace package with value unless it has a name property or the name is passed as a parameter.');
        }

        return this.replacePackageBuilder(() => value, name ?? value.name);
    }
    
    restore(moduleName: string): ITestableContainer {
        delete this._replacements[moduleName];
        delete this._repCache[moduleName];
        return this;
    }
    
    restoreAll(): ITestableContainer {
        const keys = Object.keys(this._replacements);
        let that = this;
        keys.forEach(key => that.restore(key));
        return this;
    }
    
    getModuleList(): string[] {
        const keys = Object.keys(this._registry);
        return Object.assign([], keys);
    }
}

const core = new Container();
const registry = core as IDependencyContainer;
const manager = core as IDependencyManager;
const container = core as IContainer;

export {
    registry,
    manager,
    container,
};