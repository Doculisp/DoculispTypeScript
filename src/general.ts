import { IRegisterable } from "./types.containers";
import { IFail, ILocation, ISuccess, IUtil } from "./types.general";

class Location implements ILocation {
    private readonly _line: number;
    private readonly _char: number;
    
    constructor(line: number, char: number) {
        this._line = line;
        this._char = char;
    }

    public get line(): number {
        return this._line;
    }

    public get char(): number {
        return this._char;
    }

    toString(): string {
        return `{ line: ${this._line}, char: ${this._char} }`;
    }

    asJson() {
        return { line: this._line, char: this._char };
    }
}

function buildGeneral(): IUtil {
    function ok<T>(successfulValue: T) : ISuccess<T> {
        return {
            value: successfulValue,
            success: true,
        };
    };

    function fail(message: string, documentPath: string) : IFail {
        return {
            message,
            documentPath,
            success: false,
        };
    };

    function location(line: number, char: number): ILocation {
        return new Location(line, char);
    }

    return {
        ok,
        fail,
        location,
    };
}

const astParser: IRegisterable = {
    builder: () => buildGeneral(),
    name: 'util',
    singleton: true,
    dependencies: []
};

export {
    astParser,
};