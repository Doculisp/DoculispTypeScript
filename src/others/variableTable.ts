import { IDictionary, IRegisterable } from "../types/types.containers";
import { IVariableExists, IVariableString, IVariableTable, IVariableTestable, Savable } from "../types/types.variableTable";

function buildVariableTable(): IVariableTestable {
    class VariableTable implements IVariableTable, IVariableTestable {
        private table: IDictionary<Savable> = {};
        private parent: IVariableTable | false = false;

        constructor();
        constructor(baseTable: (IVariableTable & IVariableTestable));
        constructor(baseTable?: (IVariableTable & IVariableTestable)) {
            if(!baseTable) {
                return;
            }

            const baseKeys = baseTable.getKeys();

            baseKeys.forEach(key => this.table[key] = baseTable.getValue(key) as Savable);
            this.parent = baseTable;
        }

        createChild(): IVariableTable {
            return new VariableTable(this);
        }
        
        addValue<T extends Savable>(key: string, value: T): IVariableExists {
            this.table[key] = value;
            return this;
        }

        addValueToStringList(key: string, value: IVariableString): IVariableExists {
            if(this.table[key]) {
                let tableValue = this.table[key] as Savable;
                if(tableValue.type === 'variable-array-string')
                    if(!tableValue.value.some(t => t.value == value.value)) {
                        tableValue.value.push(value);
                    }
            }
            else {
                this.table[key] = { value: [value], type: 'variable-array-string' };
            }
            return this;
        }

        private parentHasKey(key: string): boolean {
            return !!this.parent && !!this.parent.hasKey(key);
        }

        hasKey(key: string): boolean {
            return !!this.table[key] || this.parentHasKey(key);
        }

        private getValueFromParent<T extends Savable>(key: string): T | false {
            return !!this.parent ? this.parent.getValue(key) : false;
        }

        getValue<T extends Savable>(key: string): T | false {
            if(!!this.table[key]) {
                return this.table[key] as T;
            }

            return this.getValueFromParent(key);
        }

        asJson(): any {
            const keys = Object.keys(this.table);
            const ret: IDictionary<any> = {};

            keys.forEach(key => ret[key] = this.table[key]);

            return ret;
        }

        clear(): IVariableTestable {
            const keys = Object.keys(this.table);

            keys.forEach(key => delete this.table[key]);

            return this;
        }

        private addValueToParent<T extends Savable>(key: string, value: T): void {
            if(this.parent) {
                this.parent.addGlobalValue(key, value);
            }
        }

        addGlobalValue<T extends Savable>(key: string, value: T): IVariableExists {
            this.addValueToParent(key, value);
            this.addValue(key, value);
            return this;
        }

        addGlobalValueToStringList(key: string, value: IVariableString): IVariableExists {
            if(this.parent) {
                this.parent.addGlobalValueToStringList(key, value);
            }

            this.addValueToStringList(key, value);

            return this;
        }

        private getParentKeys(): string[] {
            if(!this.parent) {
                return [];
            }

            return this.parent.getKeys();
        }
        
        getKeys(): string[] {
            const mine = Object.keys(this.table);
            const theirs = this.getParentKeys();

            const both = new Set([...mine, ...theirs]);

            return Array.from(both);
        }
    }

    return new VariableTable();
}

const variableBuilder: IRegisterable = {
    builder: () => buildVariableTable(),
    name: 'variableTable',
    dependencies: [],
    singleton: true
};

export {
    variableBuilder,
};
