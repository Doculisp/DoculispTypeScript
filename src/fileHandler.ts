import { IRegisterable } from "./types.containers";
import { IFileHandler } from "./types.fileHandler";

import { IUtil, Result } from "./types.general";

function buildLoader(util: IUtil, fs: any): IFileHandler {
    function load(path: string): Result<string> {
        try {
            const value: string = fs.readFileSync(path, {encoding: 'utf8'});
            return util.ok(value);
        } catch (error) {
            return util.fail(`${error}`, path);
        }
    }

    return {
        load
    };
}

const loader: IRegisterable = {
    builder: (util: IUtil, fs: any) => buildLoader(util, fs),
    name: 'fileHandler',
    dependencies: ['util', 'fs'],
    singleton: true
};

export {
    loader,
};