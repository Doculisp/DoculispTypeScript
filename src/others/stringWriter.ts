import { AstPart, IAst, IHeader, ILoad, ISectionWriter, ITableOfContents, ITitle, IWrite } from "../types/types.ast";
import { IRegisterable } from "../types/types.containers";
import { ILocation, IUtil, Result } from "../types/types.general";
import { IStringWriter } from "../types/types.stringWriter";

class StringBuilder {
    private _lines: string[];

    constructor() {
        this._lines = [];
    }

    addLine(): StringBuilder
    addLine(value: string): StringBuilder
    addLine(value?: string): StringBuilder {
        let v = value ?? '';
        this._lines.push(v);
        return this;
    }

    add(value: string): StringBuilder {
        if(!value) {
            return this;
        }

        if(this._lines.length === 0) {
            this._lines.push('');
        }

        this._lines[this._lines.length - 1] += value;
        return this;
    }

    toString(): string {
        return this._lines.join('\n');
    }

    get length(): number {
        let i = 0;

        this._lines.forEach(l => i += l.length);

        return i + this._lines.length - 1;
    }

    get lineLength(): number {
        if(this._lines.length < 1) {
            return 0;
        }

        return this._lines[this._lines.length - 1]?.length || 0;
    }
}

function isNewLine(previousLocation: ILocation, currentLocation: ILocation): boolean {
    return !(
        previousLocation.documentPath === currentLocation.documentPath
        && previousLocation.line === currentLocation.line
    );
}

function writeAstWrite(astWrite: IWrite) : string {
    return astWrite.value;
}

function writeAstTitle(astTitle: ITitle): string {
    const sb = new StringBuilder();

    sb.add(astTitle.label);

    if(astTitle.subtitle) {
        sb.addLine();
        sb.addLine(astTitle.subtitle);
    }

    return sb.toString();
}

function writeAstHeader(astHeader: IHeader): string {
    const headMarker = '#'.repeat(astHeader.depthCount);
    return `${headMarker} ${astHeader.text} ${headMarker}`;
}

function writeTableOfContents(toc: ITableOfContents, loads: ILoad[]): string {
    function findTitle(ast: AstPart[]): ITitle | null {
        for (let index = 0; index < ast.length; index++) {
            const element = ast[index];
            if(!element) {
                continue;
            }

            if(element.type === 'ast-title') {
                return element;
            }
        }

        return null;
    }

    function useLabel(append: (text: string) => void, title: string, linkText: string, label?: string): void {
        let lblText = !!label ? `${label}: `: '';
        
        append(`[${lblText}${title}](${linkText})`);
    }

    function ignoreLabel(append: (text: string) => void, title: string, linkText: string, label?: string): void {
        return useLabel(append, title, linkText);
    }

    function writeRawLink(appender: (append: (text: string) => void, title: string, linkText: string, label?: string) => void): (sb: StringBuilder, title: string, linkText: string, label: string) => void {
        return function(sb: StringBuilder, title: string, linkText: string, label: string): void {
            appender(text => sb.addLine(text), title, linkText, label);
        }
    }

    function writeNumbered(appender: (append: (text: string) => void, title: string, linkText: string, label?: string) => void): (sb: StringBuilder, title: string, linkText: string, label: string) => void {
        let cnt = 1;
        return function(sb: StringBuilder, title: string, linkText: string, label: string): void {
            appender(text => sb.add(`${cnt}. ${text}`), title, linkText, label);
            cnt++;
        }
    }

    function writeBulleted(appender: (append: (text: string) => void, title: string, linkText: string, label?: string) => void): (sb: StringBuilder, title: string, linkText: string, label: string) => void {
        return function(sb: StringBuilder, title: string, linkText: string, label: string): void {
            appender(text => sb.add(`* ${text}`), title, linkText, label);
        }
    }

    function writeTable(loads: ILoad[], addRow: (sb: StringBuilder, title: string, linkText: string, label: string) => void): string {
        const sb = new StringBuilder();

        for (let index = 0; index < loads.length; index++) {
            const element = loads[index];
            if(!element) {
                continue;
            }
    
            if(!element.document) {
                continue;
            }
    
            const doc = element.document;
            const title = findTitle(doc.ast);
    
            if(!title) {
                continue;
            }
    
            if(0 < sb.length) {
                sb.addLine();
            }

            addRow(sb, title.title, title.ref_link, element.sectionLabel);
        }
        
        if(0 < sb.length) {
            sb.addLine();
        }
    
        return sb.toString();
    }

    switch (toc.bulletStyle) {
        case 'labeled':
            return writeTable(loads, writeRawLink(useLabel));

        case 'unlabeled':
            return writeTable(loads, writeRawLink(ignoreLabel));

        case 'numbered':
            return writeTable(loads, writeNumbered(ignoreLabel));

        case 'numbered-labeled':
            return writeTable(loads, writeNumbered(useLabel));

        case 'bulleted':
            return writeTable(loads, writeBulleted(ignoreLabel));

        case 'bulleted-labeled':
            return writeTable(loads, writeBulleted(useLabel));
    
        default:
            return `>>>> ${toc.bulletStyle} <<<<\n`;
    }
    
}

function writeContent(loads: ILoad[]): string {
    const sb = new StringBuilder();

    for (let index = 0; index < loads.length; index++) {
        const load = loads[index];
        if(!load) {
            continue;
        }

        if(!load.document) {
            continue;
        }

        const doc = load.document;
        let previous: ILocation = doc.documentOrder;

        if(0 < sb.length) {
            sb.addLine();
        }

        sb.add(writeSection(previous, doc));
    }

    return sb.toString().trim();
}

function writeSection(previous: ILocation, section: ISectionWriter): string {
    const sb = new StringBuilder();
    let previousType = '';

    for (let index = 0; index < section.ast.length; index++) {
        const ast = section.ast[index];
        if(!ast) {
            continue;
        }

        if(isNewLine(previous, ast.documentOrder)) {
            sb.addLine();
        }

        if(previousType === 'ast-write' && ast.type === 'ast-write') {
            if(previous.documentPath !== ast.documentOrder.documentPath
               || (previous.line + 2) <= ast.documentOrder.line
               || (ast.documentOrder.line + 2) <= previous.line
            ) {
                sb.addLine();
            } else if (0 < sb.lineLength) {
                //sb.add(' ');
            }
        }
        else {
            sb.addLine();
        }

        switch (ast.type) {
            case 'ast-write':
                sb.add(writeAstWrite(ast));
                break;

            case 'ast-title':
                sb.add(writeAstTitle(ast));
                break;

            case 'ast-header':
                sb.add(writeAstHeader(ast));
                break;

            case 'ast-content':
                sb.add(writeContent(section.include));
                break;

            case 'ast-toc':
                sb.add(writeTableOfContents(ast, section.include));
                break;
        
            default:
                break;
        }

        previousType = ast.type;
        previous = ast.documentOrder;
    }

    return sb.toString();
}

function buildWriter(util: IUtil) : IStringWriter {
    function writeAst(astMaybe: Result<IAst>): Result<string> {
        if(!astMaybe.success) {
            return astMaybe;
        }

        if(astMaybe.value.section.type === 'ast-empty'){
            return util.ok('');
        }

        const sb = new StringBuilder();
        const section = astMaybe.value.section;

        sb.addLine('<!-- Generated Document do not edit! -->');

        let previous: ILocation = util.location('', -1, -1, -1, -1);
        sb.addLine(writeSection(previous, section));
        
        sb.addLine();
        sb.addLine('<!-- Generated Document do not edit! -->');
        return util.ok(sb.toString());
    }

    return {
        writeAst,
    }
}

const stringWriter: IRegisterable = {
    builder: (util: IUtil) => buildWriter(util),
    name: 'stringWriter',
    dependencies: ['util'],
    singleton: false,
};

export {
    stringWriter,
};