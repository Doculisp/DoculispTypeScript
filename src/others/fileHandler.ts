import { IRegisterable } from "../types/types.containers";
import { IFileHandler } from "../types/types.fileHandler";
import { IUtil, Result } from "../types/types.general";

function buildLoader(util: IUtil, fs: any): IFileHandler {
    function load(path: string): Result<string> {
        try {
            const value: string = fs.readFileSync(path, {encoding: 'utf8'});
            return util.ok(value);
        } catch (error) {
            return util.fail(`${error}`, path);
        }
    }

    function write(path: string, text: Result<string>): Result<string> {
        if(!text.success) {
            return text;
        }

        const output = text.value;

        try {
            fs.writeFileSync(path, output, {encoding: 'utf8'});
            return text;
        }
        catch(error) {
            return util.fail(`${error}`, path);
        }
    }

    function getProcessWorkingDirectory(): string {
        return process.cwd();
    }

    function setProcessWorkingDirectory(directory: string): void {
        process.chdir(directory);
    }

    return {
        load,
        write,
        getProcessWorkingDirectory,
        setProcessWorkingDirectory,
    };
}

const loader: IRegisterable = {
    builder: (util: IUtil, fs: any) => buildLoader(util, fs),
    name: 'fileHandler',
    dependencies: ['util', 'fs'],
    singleton: true
};

export {
    loader,
};