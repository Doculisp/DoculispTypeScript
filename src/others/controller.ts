import { IAstBuilder } from "../types/types.astBuilder";
import { IRegisterable } from "../types/types.containers";
import { IController } from "../types/types.controller";
import { IFileWriter } from "../types/types.fileHandler";
import { IUtil, Result } from "../types/types.general";
import { IStringWriter } from "../types/types.stringWriter";

function buildLoader(util: IUtil, handler: IFileWriter, astBuilder: IAstBuilder, stringWrter: IStringWriter): IController {
    function _compile(sourcePath: string, destinationPath: string | false): Result<string | false> {
        const ast = astBuilder.parse(sourcePath);
        const document = stringWrter.writeAst(ast);

        if(!!destinationPath){
            const result = handler.write(destinationPath, document);

            if(!result.success) {
                return result;
            }
        } else if(!document.success) {
            return document;
        }

        return util.ok(destinationPath);
    }

    function compile(sourcePath: string, destinationPath: string): Result<string> {
        return _compile(sourcePath, destinationPath) as Result<string>;
    }

    function test(sourcePath: string): Result<false> {
        return _compile(sourcePath, false) as Result<false>;
    }

    return {
        compile,
        test,
    };
}

const controllerBuilder: IRegisterable = {
    builder: (util: IUtil, handler: IFileWriter, astBuilder: IAstBuilder, stringWrter: IStringWriter) => buildLoader(util, handler, astBuilder, stringWrter),
    name: 'controller',
    dependencies: ['util', 'fileHandler', 'astBuilder', 'stringWriter'],
    singleton: true
};

export {
    controllerBuilder,
};
