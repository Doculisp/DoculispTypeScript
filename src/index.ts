#! /usr/bin/env node
import { container } from './container';
import { Command } from 'commander';
import { Result } from './types/types.general';
import { IController } from './types/types.controller';
import { IVersion } from './types/types.version';
import figlet from 'figlet';

const program = new Command();
const controller = container.buildAs<IController>('controller');
const versionGetter = container.buildAs<IVersion>('version');

const versionMaybe = versionGetter.getVersion();


let version = '?.?.?'

if(!versionMaybe.success) {
    console.log('doculisp cannot determine its version')
} else {
    version = versionMaybe.value;
}

const helptext = '\n\n' + figlet.textSync('doculisp', {
    font: 'Cybermedium'
}) + `\n                     Version: ${version}`;

program.addHelpText('beforeAll', helptext);
program.addHelpText('afterAll', '\n\n');

program
    .name('doculisp')
    .description('A compiler for markdown')
    .version(version)
    .requiredOption('-s, --source <source_path>', 'the source file to compile')
    .requiredOption('-d, --output <output_path>', 'the output document path for the compiled markdown')
    .option('-t, --test', 'runs the compiler without generating the output file.')
    .parse(process.argv);

function reportResult(result: Result<string | false>) {
    if (result.success) {
        const message = !!result.value ? `Ok! ${result.value} successfully written` : 'Ok!';
        console.log(message);
    }
    else {
        console.log(JSON.stringify(result, null, 4));
    }
}

const options = program.opts();

const sourcePath: string = options['source'];

if (options['test']) {
    const result = controller.test(sourcePath);
    reportResult(result);
}
else {
    const outputPath: string = options['output'];
    const result = controller.compile(sourcePath, outputPath);
    reportResult(result);
}
