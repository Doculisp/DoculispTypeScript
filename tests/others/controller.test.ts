import { IDictionary, ITestableContainer } from "../../src/types/types.containers";
import { IFileWriter } from "../../src/types/types.fileHandler";
import { IPath, PathConstructor } from "../../src/types/types.filePath";
import { IUtil, Result } from "../../src/types/types.general";
import { IIncludeBuilder } from "../../src/types/types.includeBuilder";
import { IDoculisp, IEmptyDoculisp } from "../../src/types/types.astDoculisp";
import { IVariableRetriever, IVariableSaver } from "../../src/types/types.variableTable";
import { IStringWriter } from "../../src/types/types.stringWriter";
import { IController } from "../../src/types/types.controller";

import { container } from "../../src/container";
import { Options } from "approvals/lib/Core/Options";
import { configure } from "approvals/lib/config";
import { getVerifier } from "../tools";
import path from "path";
import { IProjectDocuments } from "../../src/types/types.astProject";

type FileConfig = {
    outputPath?: IPath | undefined;
    fileText?: Result<string> | undefined;
    result?: Result<string> | undefined;
};

type IncludeConfig = {
    sourcePath?: IPath | undefined;
    doculisp? : Result<IDoculisp | IEmptyDoculisp> | undefined;
    result?: Result<IDoculisp | IEmptyDoculisp> | undefined;
};

type WriterConfig = {
    astMaybe?: Result<IDoculisp | IEmptyDoculisp> | undefined;
    result?: Result<string> | undefined;
};

describe('controller', () => {
    const environment: ITestableContainer = container as ITestableContainer;
    
    let verifyAsJson: (data: any, options?: Options) => void;
    let testable: ITestableContainer = null as any;
    let util: IUtil;

    let fileConfig: FileConfig = undefined as any;
    let includeConfig: IncludeConfig = undefined as any;
    let writerConfig: WriterConfig = undefined as any;

    let pathConstructor: PathConstructor = undefined as any;
    let sut: IController = undefined as any;

    beforeAll(() => {
        verifyAsJson = getVerifier(configure);
    });
    
    function getTestResult(finalResult: any = {}) {
        const result: IDictionary<any> = {};
        let cnt = 1;
        function add(name: string, value: any) {
            if(0 < Object.keys(value).length) {
                const index = `0${cnt} ${name}`;
                result[index] = value;
                cnt++;
            }
        }

        add('Parse Ast', includeConfig);
        add('Convert to Markdown', writerConfig);
        add('Write to File', fileConfig);
        add('Final Result', finalResult);

        return result;
    }

    beforeEach(() =>{
        testable = environment.buildTestable();
        util = testable.buildAs<IUtil>('util');

        fileConfig = {};
        const fileWriter : IFileWriter = {
            write: function (path: IPath, text: Result<string>): Result<string> {
                fileConfig.outputPath = path;
                fileConfig.fileText = text;

                if(!text.success) {
                    return text;
                }

                return fileConfig.result ?? util.ok(path.fullName);
            }
        };
        testable.replaceValue(fileWriter, 'fileHandler')

        includeConfig = {};
        const emptyResult: IEmptyDoculisp = {
            type: "doculisp-empty"
        };

        const includeBuilder: IIncludeBuilder = {
            parse: function (path: IPath, _variableTable: IVariableSaver): Result<IDoculisp | IEmptyDoculisp> {
                includeConfig.sourcePath = path;
                return includeConfig.result ?? util.ok(emptyResult);
            },
            parseExternals: function (doculisp: Result<IDoculisp | IEmptyDoculisp>, _variableTable: IVariableSaver): Result<IDoculisp | IEmptyDoculisp> {
                includeConfig.doculisp = doculisp;
                if (!doculisp.success) {
                    return doculisp;
                }

                return includeConfig.result ?? util.ok(emptyResult);
            },
            parseProject: function (path: IPath): Result<IProjectDocuments> {
                throw new Error("parseProject intentionally not implemented.");
            }
        };
        testable.replaceValue(includeBuilder, 'includeBuilder')

        writerConfig = {};
        const stringWriter: IStringWriter = {
            writeAst: function (astMaybe: Result<IDoculisp | IEmptyDoculisp>, _variableTable: IVariableRetriever): Result<string> {
                writerConfig.astMaybe = astMaybe;
                if(!astMaybe.success) {
                    return astMaybe;
                }
                return writerConfig.result ?? util.ok('# Good Document #\n\nHello');
            }
        };
        testable.replaceValue(stringWriter, 'stringWriter');

        pathConstructor = function(pathString: string): IPath {
            const t: IPath = {
                extension: path.extname(pathString),
                getContainingDir: function (): IPath {
                    return pathConstructor("./");
                },
                getRelativeFrom: function (rootPath: IPath): string {
                    return "./"
                },
                type: "path",
                toJSON: function () {
                    return pathString;
                },
                fullName: pathString,
            };

            return t;
        };
        testable.replaceValue(pathConstructor, 'pathConstructor');

        sut = testable.buildAs<IController>('controller');
    });

    describe('test', () => {
        it('should test handle a successful file', () => {
            const sourcePath = pathConstructor('./someFile.md');
            sut.test(sourcePath);
    
            verifyAsJson(getTestResult());
        });
    
        it('should fail a file that cannot parse an ast', () => {
            const sourcePath = pathConstructor('./someFile.md');
            includeConfig.result = util.fail('A bad parse', sourcePath);
            const result = sut.test(sourcePath);
    
            verifyAsJson(getTestResult(result));
        });
    
        it('should fail a file that cannot be converted to markdown', () => {
            const sourcePath = pathConstructor('./someFile.md');
            writerConfig.result = util.fail('Unable to write', sourcePath);
            const result = sut.test(sourcePath);
    
            verifyAsJson(getTestResult(result));
        });
    });

    describe('compile', () => {
        it('should be successful if everything is successful', () => {
            const sourcePath = pathConstructor('./someFile.md');
            const destinationPath = pathConstructor('./README.md');
            sut.compile(sourcePath, destinationPath);
    
            verifyAsJson(getTestResult());
        });

        it('should fail if a file cannot parse an ast', () => {
            const sourcePath = pathConstructor('./someFile.md');
            const destinationPath = pathConstructor('./README.md');
            includeConfig.result = util.fail('Unable to parse ast', sourcePath);

            const result = sut.compile(sourcePath, destinationPath);

            verifyAsJson(getTestResult(result));
        });
    
        it('should fail a file that cannot be converted to markdown', () => {
            const sourcePath = pathConstructor('./someFile.md');
            const destinationPath = pathConstructor('./README.md');
            writerConfig.result = util.fail('Unable to write', sourcePath);
            
            const result = sut.compile(sourcePath, destinationPath);
    
            verifyAsJson(getTestResult(result));
        });

        it('should fail if writing the file fails', () => {
            const sourcePath = pathConstructor('./someFile.md');
            const destinationPath = pathConstructor('./README.md');
            fileConfig.result = util.fail('Unable to write file', destinationPath);

            const result = sut.compile(sourcePath, destinationPath);

            verifyAsJson(getTestResult(result));
        });
    });
});