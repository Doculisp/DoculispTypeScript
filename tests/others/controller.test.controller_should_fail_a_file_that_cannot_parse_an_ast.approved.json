{
  "fileConfig": {},
  "finalResult": {
    "documentPath": "./someFile.md",
    "message": "A bad parse",
    "success": false
  },
  "includeConfig": {
    "parseResult": {
      "documentPath": "./someFile.md",
      "message": "A bad parse",
      "success": false
    },
    "sourcePath": "./someFile.md"
  },
  "writerConfig": {
    "astMaybe": {
      "documentPath": "./someFile.md",
      "message": "A bad parse",
      "success": false
    }
  }
}
